apps/gpc-verifier/app/util/verifyProof.ts

```diff
@@ -0,0 +1,107 @@
+import { Dispatch } from "react";
+import {
+  GPCBoundConfig,
+  GPCProof,
+  gpcArtifactDownloadURL,
+  gpcVerify,
+  deserializeGPCBoundConfig,
+  deserializeGPCRevealedClaims,
+  PODMembershipLists
+} from "@pcd/gpc";
+import { tryRecordNullifierHash } from "@/util/persistence";
+
+export const verifyProof = async (
+  boundConfig: GPCBoundConfig,
+  membershipLists: PODMembershipLists,
+  externalNullifier: string,
+  watermark: string,
+  proofStr: string,
+  setVerified: Dispatch<boolean>
+) => {
+  try {
+    if (!proofStr) {
+      throw new Error("Proof cannot be empty!");
+    }
+
+    const proofObj = JSON.parse(proofStr);
+    const vProof = JSON.parse(proofObj.proof) as GPCProof;
+
+    // Use the boundConfig we provided.
```

Comment:
This is tricky and needs more explanation.  I actually like it as a real-world approach, since it's forcing the verification compiler to do your checking for you.  We just need to explain what's going on.

Calling `boundConfig` something like `requestorConfig` or `originalConfig` may help, but then I think there needs to be a comment explaining why it's important to do this, so the prover can't present a proof something different from what you asked them to prove.

An alternative way of expressing this which is more intuitive is to check that the proofconfig matches the boundconfig via a deep comparision, after removing the circuit identifier from both.

The complexity of doing this right (which I've also been reviewing for Rob in ZuAuth) has made me realize we need a first-class concept of a ProofRequest which can be fed to the verifier to check all this automatically.

Reply:
+1 on have a first-class concept of a proofRequest.